### PR TITLE
Prefer coins that have fewer ancestors, sanity check txn before ATMP

### PR DESCRIPTION
**Pull request by instagibbs**
This does a few extra possible runs of SelectCoinsMinConf, each with larger allowable mempool ancestor numbers, up to the acceptable mempool limit. As long as each input is not above this value, it passes. This means the sum of the history could actually be larger.

This is why I then catch the transaction at the end of CreateTransaction, and directly check. It simply fails, and does not retry. This logic is gated by a new command line argument `-rejectlongwalletchains` to regain previous behavior. We can most likely remove this check once we have better rebroadcasting systems.

### CODE REVIEW

**Comment by sipa** [comment]
There should probably be a function in the place where mempool acceptance is decided (main.cpp now, validation.cpp after The Main Split) to do preliminary checking like thid.

**Comment by gmaxwell** [comment]
Probably not for now but when I read this code I wondered if we need an "available balance"-- e.g. the sum of the coins that selectcoins would consider using... so that transaction failures due to this (and spendunconfirmed set off) are more explicable.

**Comment by gmaxwell** [comment]
FWIW, I think this should be backported.

**Comment by instagibbs** [comment]
While writing tests I found that this logic somehow is falling through on non-default ancestorcount limit. Debugging.

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -1957,6 +1957,20 @@ void CWallet::AvailableCoins(vector<COutput>& vCoins, bool fOnlyConfirmed, const
             if (nDepth == 0 && !pcoin->InMempool())
                 continue;
 
+            LockPoints lp;
+            CTxMemPoolEntry entry(*pcoin, 0, 0, 0, 0, false, 0, false, 0, lp);
```
Comment (sdaftuar) [inline src/wallet/wallet.cpp]: Since we've already checked that pcoin is in the mempool, we don't need to do any of this -- we can just use the cached values.  (I think the only way the calculation you do below can fail is if the limits are violated during a reorg.)

**Comment by instagibbs** [comment]
Updated the pull request to new design discussed on IRC.

File: src/wallet/wallet.h [hunk 2]
```diff
@@ -400,6 +400,7 @@ class CWalletTx : public CMerkleTx
     bool IsEquivalentTo(const CWalletTx& tx) const;
 
     bool InMempool() const;
+    uint64_t GetMempoolAncestorCount() const;
```
Comment (sdaftuar) [inline src/wallet/wallet.h]: nit: add documentation to this function that the tx must be in the mempool in order to call?

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -2042,6 +2048,9 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, int nConfMine, int
         if (output.nDepth < (pcoin->IsFromMe(ISMINE_ALL) ? nConfMine : nConfTheirs))
             continue;
 
+        if (pcoin->InMempool() && pcoin->GetMempoolAncestorCount() > nMaxAncestors)
```
Comment (sdaftuar) [inline src/wallet/wallet.cpp]: Perhaps worth a comment here that we're assuming `AvailableCoins()` has already filtered out any unconfirmed and not-in-mempool coins?

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -2558,6 +2558,21 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
         }
     }
 
+    // Lastly, ensure it doesn't have too many ancestors
```
Comment (sdaftuar) [inline src/wallet/wallet.cpp]: nit: perhaps this comment should be "Lastly, ensure this tx will pass the mempool's chain limits"

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -2558,6 +2558,21 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
...
+    LockPoints lp;
+    CTxMemPoolEntry entry(txNew, 0, 0, 0, 0, false, 0, false, 0, lp);
+    CTxMemPool::setEntries setAncestors;
+    size_t nLimitAncestors = GetArg("-limitancestorcount", DEFAULT_ANCESTOR_LIMIT);
+    size_t nLimitAncestorSize = GetArg("-limitancestorsize", DEFAULT_ANCESTOR_SIZE_LIMIT)*1000;
+    size_t nLimitDescendants = GetArg("-limitdescendantcount", DEFAULT_DESCENDANT_LIMIT);
+    size_t nLimitDescendantSize = GetArg("-limitdescendantsize", DEFAULT_DESCENDANT_SIZE_LIMIT)*1000;
+    std::string errString;
+    if (!mempool.CalculateMemPoolAncestors(entry, setAncestors, nLimitAncestors, nLimitAncestorSize, nLimitDescendants, nLimitDescendantSize, errString)) {
+        strFailReason = _("Transaction has too long of a mempool chain");
```
Comment (sdaftuar) [inline src/wallet/wallet.cpp]: The errString filled in by CMPA might be useful for debugging issues (eg a LogPrintf() so that someone debugging could inspect), though i see we don't log any info about the transaction we tried to generate, so maybe not worth it...

**Review by sdaftuar — COMMENTED** [review]
Code review ack, just a couple nits.  Will test.

**Comment by instagibbs** [comment]
I also think this is good for backport.

File: src/wallet/wallet.cpp [hunk 6]
```diff
@@ -2042,6 +2048,9 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, int nConfMine, int
         if (output.nDepth < (pcoin->IsFromMe(ISMINE_ALL) ? nConfMine : nConfTheirs))
             continue;
 
+        if (pcoin->GetMempoolAncestorCount() > nMaxAncestors)
```
Comment (sdaftuar) [inline src/wallet/wallet.cpp]: I think you need the pcoin->InMempool() check you had before, in case you're looking at in-chain coins?

**Comment by gmaxwell** [comment]
Looks good! -- so, I think the additional test in "Don't return success"  should be controlled by an option, since the failure is only temporary (at least once we fix the rebroadcast bug) and many callers handle failure to create a transaction poorly. I might even go as far as to suggest that this option be set in 0.13.2 to continue to create the transaction because that would be less of an API change.  Perhaps @sdaftuar or @laanwj has an opinion on that?

Other than that, utACK.

**Comment by instagibbs** [comment]
Couple of fixes/changes:

>if (pcoin->InMempool() && pcoin->GetMempoolAncestorCount() > nMaxAncestors)

is now

>if (pcoin->InMempool() && pcoin->GetMempoolAncestorCount() >= nMaxAncestors)

so it actually filters correctly.

I also revamped the tests to catch the various cases.

@gmaxwell we'll have to catch that logic in two places, SelectCoinsMinConf(at the last instance of it) and at the end of CreateTransaction. I suppose once rebroadcasting is fixed, we can set the default to "false" and let it take care of it. I'll wait to see what others are thinking. I think current behavior is horrendous :)

**Comment by gmaxwell** [comment]
Current behavior is horrendous indeed. Thank you for working on this.

**Comment by sdaftuar** [comment]
@gmaxwell No strong opinion on the issue of default behavior for 0.13.2, but given that you currently have to restart your node in order to rebroadcast, I'd lean towards making the default behavior be to enforce this new restriction, rather than create a tx that doesn't make it to the mempool.

File: src/wallet/wallet.cpp [hunk 7]
```diff
@@ -2042,6 +2048,9 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, int nConfMine, int
         if (output.nDepth < (pcoin->IsFromMe(ISMINE_ALL) ? nConfMine : nConfTheirs))
             continue;
 
+        if (pcoin->InMempool() && pcoin->GetMempoolAncestorCount() >= nMaxAncestors)
```
Comment (sdaftuar) [inline src/wallet/wallet.cpp]: @morcos and I were discussing, I think this would make more sense to limit based on both ancestor and descendant count, something like:

```
if (pcoin->InMempool() && max(pcoin->GetMempoolAncestorCount(), pcoin->GetMempoolDescendantCount()) >= nMaxChain)
    continue;
````

Otherwise, if you have an unspent output with 1 ancestor but 25 descendants, you (needlessly) could fail on one of the early calls.

File: src/wallet/wallet.cpp [hunk 8]
```diff
@@ -2167,10 +2176,14 @@ bool CWallet::SelectCoins(const vector<COutput>& vAvailableCoins, const CAmount&
...
+
     bool res = nTargetValue <= nValueFromPresetInputs ||
-        SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 1, 6, vCoins, setCoinsRet, nValueRet) ||
-        SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 1, 1, vCoins, setCoinsRet, nValueRet) ||
-        (bSpendZeroConfChange && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 0, 1, vCoins, setCoinsRet, nValueRet));
+        SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 1, 6, 0, vCoins, setCoinsRet, nValueRet) ||
+        SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 1, 1, 0, vCoins, setCoinsRet, nValueRet) ||
+        (bSpendZeroConfChange && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 0, 1, nLimitAncestors/4, vCoins, setCoinsRet, nValueRet)) ||
+        (bSpendZeroConfChange && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 0, 1, nLimitAncestors/2, vCoins, setCoinsRet, nValueRet)) ||
+        (bSpendZeroConfChange && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 0, 1, nLimitAncestors, vCoins, setCoinsRet, nValueRet));
```
Comment (sdaftuar) [inline src/wallet/wallet.cpp]: Instead of N/4, N/2, and N for the ancestor/chain limit passed in to SCMC, how about: 2, 4, N/2, N?  (And set N = min(limitancestorcount, limitdescendantcount), assuming we go with ancestor and descendant limiting as I mentioned in another comment.)  This way we first try utxo's from transactions that have no ancestors or descendants, which should provide a strong preference for not creating long chains.

**Comment by instagibbs** [comment]
@sdaftuar made sense to me, done

**Comment by laanwj** [comment]
>  but given that you currently have to restart your node in order to rebroadcast

Why is that so? You can't take the tx hex and send it through `sendrawtransaction` in this case?

File: src/wallet/wallet.h [hunk 9]
```diff
@@ -685,7 +688,7 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
      * completion the coin set and corresponding actual target value is
      * assembled
      */
-    bool SelectCoinsMinConf(const CAmount& nTargetValue, int nConfMine, int nConfTheirs, std::vector<COutput> vCoins, std::set<std::pair<const CWalletTx*,unsigned int> >& setCoinsRet, CAmount& nValueRet) const;
+    bool SelectCoinsMinConf(const CAmount& nTargetValue, const int nConfMine, const int nConfTheirs, const uint64_t nMaxAncestors, std::vector<COutput> vCoins, std::set<std::pair<const CWalletTx*,unsigned int> >& setCoinsRet, CAmount& nValueRet) const;
```
Comment (sipa) [inline src/wallet/wallet.h]: Specifying const for a pass-by-value argument has no effect in the declaration. You can add const in the function definition if you want to protect yourself from accidentally modifying the variable even when the definition doesn't have const.

File: src/wallet/wallet.cpp [hunk 10]
```diff
@@ -2167,10 +2182,15 @@ bool CWallet::SelectCoins(const vector<COutput>& vAvailableCoins, const CAmount&
...
 
+    size_t nLimitAncestors = std::min(GetArg("-limitancestorcount", DEFAULT_ANCESTOR_LIMIT), GetArg("-limitdescendantcount", DEFAULT_DESCENDANT_LIMIT));
+
     bool res = nTargetValue <= nValueFromPresetInputs ||
-        SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 1, 6, vCoins, setCoinsRet, nValueRet) ||
-        SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 1, 1, vCoins, setCoinsRet, nValueRet) ||
-        (bSpendZeroConfChange && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 0, 1, vCoins, setCoinsRet, nValueRet));
+        SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 1, 6, 0, vCoins, setCoinsRet, nValueRet) ||
+        SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 1, 1, 0, vCoins, setCoinsRet, nValueRet) ||
+        (bSpendZeroConfChange && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 0, 1, 2, vCoins, setCoinsRet, nValueRet)) ||
```
Comment (sipa) [inline src/wallet/wallet.cpp]: What if nLimitAncestors is less than 4?

File: src/wallet/wallet.cpp [hunk 11]
```diff
@@ -2565,6 +2565,20 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
         }
     }
 
+    // Lastly, ensure this tx will pass the mempool's chain limits
```
Comment (sipa) [inline src/wallet/wallet.cpp]: Can you abstract this fragment out, and turn it into a CTxMempool method?

**Comment by gmaxwell** [comment]
> but given that you currently have to restart your node in order to rebroadcast

How about we fix that? (#9290)

File: src/wallet/wallet.cpp (hunk 10 above)
Comment (gmaxwell) [inline src/wallet/wallet.cpp]: It calls Selectmincoins redundantly a bit... this doesn't seem like a big deal to me.

**Comment by sdaftuar** [comment]
@laanwj Regarding rebroadcast, I meant without doing something herculean -- I think with current behavior you'd have to first find the txid that was committed but then not broadcast (since the RPC call will just return failure), and then rebroadcast it using something like sendrawtransaction.  But anyway it should be moot after #9290, which I think makes sense.

**Comment by sdaftuar** [comment]
Moderately-tested ACK 2da773454c02990665ecd1b03f60a6ab09bdf22d

I tried to test this a couple different ways.  One simple demonstration of this PR's benefit was in a situation where you have 2 (confirmed) UTXO's in your wallet, and you're sending lots of small valued transactions to addresses outside your wallet.  Before this PR, the test would consistently fail at the 26th spend (with the ensuing chaos of the rpc call failure, while the transaction was still committed).  After this PR, I'm able to consistently create 50 transactions before failure, which is optimal.

File: qa/rpc-tests/wallet.py [hunk 12]
```diff
@@ -346,5 +348,60 @@ def run_test (self):
...
+            self.nodes[0].lockunspent(False, [{"txid":txids[(i+1)%2], "vout":0}])
+            amount = self.nodes[0].listunspent(0, 0, [chain_addrs[i]])[0]["amount"]
+            try:
+                txids[i] = self.nodes[0].sendtoaddress(chain_addrs[i], amount, "", "", True)
+                assert(False)
+            except JSONRPCException as e:
+                assert("Insufficient funds" in e.error['message'])
+            self.nodes[0].lockunspent(True, [{"txid":txids[(i+1)%2], "vout":0}])
+
+        assert_equal(node0_balance, self.nodes[0].getbalance("*", 0))
```
Comment (morcos) [inline qa/rpc-tests/wallet.py]: Overall I find this test a bit hard to follow.  I wonder if you could make it a bit clearer.

It does a good job of testing that the send failures don't leave funds unavailable (b/c the transaction isn't committed), but I'm not sure that's even what @gmaxwell wants, although I do think it is a good idea.   But it doesn't test the hopefully more optimal behavior of being less likely to create these long chains.

Also, please , please, please, please don't use getbalance("*"), you can use getbalance() + getunconfirmedbalance() if you need the total including mempool transactions...

**Review by morcos — APPROVED** [review]
code review ACK modulo nits about the test

File: qa/rpc-tests/wallet.py (hunk 12 above)
Comment (instagibbs) [inline qa/rpc-tests/wallet.py]: Particularly I was interested in checking it errors out when, and how. So I build two long chains, start checking that sending will fail due to final checks when combining the chains, not "lack of coins", then once it gets long enough, both coins are already too long to include.

This test might be too obscure/specific, but it helped convince me that the code is doing what I wanted it to do. I should probably add a simple test like @sdaftuar 's where it shows it preferentially choosing shorter chained outputs first.

>please don't use getbalance("*")

Leftover, sorry. Will fix.

**Comment by instagibbs** [comment]
Added init flag `-rejectlongwalletchains` which is by default `True`. When set to false only the preferential SelectCoinsMinConf behavior remains, with the last iteration accepting any length.

File: src/wallet/wallet.cpp [hunk 13]
```diff
@@ -2167,10 +2182,16 @@ bool CWallet::SelectCoins(const vector<COutput>& vAvailableCoins, const CAmount&
...
+    bool frejectLongChains = GetBoolArg("-rejectlongwalletchains", DEFAULT_REJECT_TOOLONG_CHAINS);
+
     bool res = nTargetValue <= nValueFromPresetInputs ||
-        SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 1, 6, vCoins, setCoinsRet, nValueRet) ||
-        SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 1, 1, vCoins, setCoinsRet, nValueRet) ||
-        (bSpendZeroConfChange && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 0, 1, vCoins, setCoinsRet, nValueRet));
+        SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 1, 6, 0, vCoins, setCoinsRet, nValueRet) ||
+        SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 1, 1, 0, vCoins, setCoinsRet, nValueRet) ||
+        (bSpendZeroConfChange && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 0, 1, 2, vCoins, setCoinsRet, nValueRet)) ||
+        (bSpendZeroConfChange && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 0, 1, 4, vCoins, setCoinsRet, nValueRet)) ||
```
Comment (gmaxwell) [inline src/wallet/wallet.cpp]: So now when nLimitAncestors is set to 3 it will violate it here?

Comment (instagibbs) [inline src/wallet/wallet.cpp]: Perhaps just pick larger denominators? n/10, n/5, n/2, n(or inf).

Comment (sdaftuar) [inline src/wallet/wallet.cpp]: @gmaxwell We're already not guaranteed to find an optimal spend, this is just best-efforts anyway, so in my opinion it's not worth trying to optimize further for users using unusual values of these policy limits, which we already discourage changes to by hiding away in help-debug.

At any rate I think having explicit small values here is overall a win, in case someone were using non-default arguments for nLimitAncestors -- this ensures that the coin selection is always first going to try coins that have no ancestors/descendants.

Comment (gmaxwell) [inline src/wallet/wallet.cpp]: 0, (N+1)/4, (N+1)/2, N (or inf)   gives all unique results for everything except 1 and 2... so the user would waste some computation for 1,2.

Comment (gmaxwell) [inline src/wallet/wallet.cpp]: Restating my suggestion:  It's better to have more small number attempts because the large attempts will intermix chains and trash your short chains.  So  min(2,limit/4), min(4, limit/3), limit/2, limit would be a better pick... for example

Comment (morcos) [inline src/wallet/wallet.cpp]: ok sounds good to me

Comment (instagibbs) [inline src/wallet/wallet.cpp]: done

**Comment by gmaxwell** [comment]
This is the prettiest blue bike-shed ever.  ACK.

**Comment by sdaftuar** [comment]
Note that if someone uses `-rejectlongwalletchains=0`, then in my simple use-case above (2 confirmed UTXOs that you're repeatedly making small spends from) this current code will unnecessarily produce a too-long-chain, because we don't try calling SCMC with nlimitancestors.

After #9290 and #9302 I think this won't be a big deal, so given that we're defaulting this to true I don't think it matters much, and this shed has been repainted so many times we're probably running out of paint.  re-ACK cb6a6be.

File: src/wallet/wallet.cpp [hunk 14]
```diff
@@ -2042,6 +2054,9 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, int nConfMine, int
         if (output.nDepth < (pcoin->IsFromMe(ISMINE_ALL) ? nConfMine : nConfTheirs))
             continue;
 
+        if (pcoin->InMempool() && std::max(pcoin->GetMempoolAncestorCount(), pcoin->GetMempoolDescendantCount()) >= nMaxAncestors)
```
Comment (luke-jr) [inline src/wallet/wallet.cpp]: Why would we compare the descendant count to the max *ancestors*? Shouldn't it be compared to max descendants?

File: src/wallet/wallet.cpp [hunk 15]
```diff
@@ -2167,10 +2182,16 @@ bool CWallet::SelectCoins(const vector<COutput>& vAvailableCoins, const CAmount&
             ++it;
     }
 
+    size_t nLimitAncestors = std::min(GetArg("-limitancestorcount", DEFAULT_ANCESTOR_LIMIT), GetArg("-limitdescendantcount", DEFAULT_DESCENDANT_LIMIT));
+    bool frejectLongChains = GetBoolArg("-rejectlongwalletchains", DEFAULT_REJECT_TOOLONG_CHAINS);
```
Comment (luke-jr) [inline src/wallet/wallet.cpp]: Probably should capitalise the 'R' in the var name.

File: src/wallet/wallet.cpp [hunk 16]
```diff
@@ -2167,10 +2182,16 @@ bool CWallet::SelectCoins(const vector<COutput>& vAvailableCoins, const CAmount&
...
-        SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 1, 1, vCoins, setCoinsRet, nValueRet) ||
-        (bSpendZeroConfChange && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 0, 1, vCoins, setCoinsRet, nValueRet));
+        SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 1, 6, 0, vCoins, setCoinsRet, nValueRet) ||
+        SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 1, 1, 0, vCoins, setCoinsRet, nValueRet) ||
+        (bSpendZeroConfChange && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 0, 1, std::min((size_t)2, nLimitAncestors/4), vCoins, setCoinsRet, nValueRet)) ||
+        (bSpendZeroConfChange && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 0, 1, std::min((size_t)4, nLimitAncestors/3), vCoins, setCoinsRet, nValueRet)) ||
+        (bSpendZeroConfChange && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 0, 1, nLimitAncestors/2, vCoins, setCoinsRet, nValueRet)) ||
+        (bSpendZeroConfChange && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, 0, 1, frejectLongChains ? nLimitAncestors : std::numeric_limits<uint64_t>::max(), vCoins, setCoinsRet, nValueRet));
```
Comment (luke-jr) [inline src/wallet/wallet.cpp]: Shouldn't `nLimitAncestors` here be `(nLimitAncestors - 1)`, because otherwise the new transaction would be 1 beyond the limit?

Comment (luke-jr) [inline src/wallet/wallet.cpp]: Even if `frejectLongChains` is set, shouldn't we prefer coins <nLimitAncestors when possible?

File: src/wallet/wallet.cpp (hunk 15 above)
Comment (luke-jr) [inline src/wallet/wallet.cpp]: Might be best to have the default const, variable, and argument be more closely matched.

Suggest `DEFAULT_WALLET_REJECT_LONG_CHAINS`, `fRejectLongChains`, and `-walletrejectlongchains`

File: src/wallet/wallet.cpp [hunk 17]
```diff
@@ -2545,6 +2566,21 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
...
 
+    if (GetBoolArg("-rejectlongwalletchains", DEFAULT_REJECT_TOOLONG_CHAINS)) {
+        // Lastly, ensure this tx will pass the mempool's chain limits
+        LockPoints lp;
+        CTxMemPoolEntry entry(txNew, 0, 0, 0, 0, false, 0, false, 0, lp);
+        CTxMemPool::setEntries setAncestors;
+        size_t nLimitAncestors = GetArg("-limitancestorcount", DEFAULT_ANCESTOR_LIMIT);
+        size_t nLimitAncestorSize = GetArg("-limitancestorsize", DEFAULT_ANCESTOR_SIZE_LIMIT)*1000;
+        size_t nLimitDescendants = GetArg("-limitdescendantcount", DEFAULT_DESCENDANT_LIMIT);
+        size_t nLimitDescendantSize = GetArg("-limitdescendantsize", DEFAULT_DESCENDANT_SIZE_LIMIT)*1000;
```
Comment (luke-jr) [inline src/wallet/wallet.cpp]: Might make sense to make a `bool CalculateMemPoolAncestors(const CTxMemPoolEntry &entry, setEntries &setAncestors, std::string &errString, bool fSearchForParents = true) const;` for this common code.

OTOH, we should probably split the policy logic out of ATMP, so this refactoring may not make sense long-term.

File: src/wallet/wallet.cpp (hunk 14 above)
Comment (instagibbs) [inline src/wallet/wallet.cpp]: misnomer, nMaxAncestors is the minimum of the two parameters. Will fix.

File: src/wallet/wallet.cpp (hunk 16 above)
Comment (instagibbs) [inline src/wallet/wallet.cpp]: the count counts the coin itself in the chain length, so I think it's right as-is. `>=` is the check to reject.

**Comment by instagibbs** [comment]
Anticipating other changes to wallet https://github.com/bitcoin/bitcoin/pull/9302 https://github.com/bitcoin/bitcoin/pull/9290 , I changed the default for rejecting wallet transactions to false, and took the opportunity to put on a couple more layers of paint with renaming suggestions by @luke-jr , made a much simpler test that just checks for wallet coin preferential treatment, always have the wallet attempt an at-limit coin selection and then followup with an anything goes attempt.

**Comment by gmaxwell** [comment]
@instagibbs Test fails-- otherwise looking fine to me.

**Comment by instagibbs** [comment]
forgot to rename the param in description, fixed

File: src/wallet/wallet.cpp [hunk 18]
```diff
@@ -3347,6 +3384,7 @@ std::string CWallet::GetWalletHelpString(bool showDebug)
                                                             CURRENCY_UNIT, FormatMoney(DEFAULT_TRANSACTION_MINFEE)));
     strUsage += HelpMessageOpt("-paytxfee=<amt>", strprintf(_("Fee (in %s/kB) to add to transactions you send (default: %s)"),
                                                             CURRENCY_UNIT, FormatMoney(payTxFee.GetFeePerK())));
+    strUsage += HelpMessageOpt("-walletrejectlongchains", strprintf(_("Wallet will not create transactions that violate mempool chain limits (default: %u"), DEFAULT_WALLET_REJECT_LONG_CHAINS));
```
Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: I don't think it makes sense to expose such a detail to every user. (Note that all the `-limit<whatever>` options are only displayed to advanced users.)

Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: In light of search for a general solution to the problem, you might want to consider introducing an option that makes the wallet ignore whatever mempool policies are in place right now.

When this option is enabled, we get the current behavior, i.e. transactions might not be accepted to the mempool and might not be relayed. However, when the option is disabled and the mempool policy is considered during transaction creation, every transaction that is created also is accepted to the mempool. If no such transaction can be created, the call fails.

**Comment by MarcoFalke** [comment]
Concept ACK

**Comment by morcos** [comment]
Tested ACK 63c17b8

However I still don't really love the test.  For instance the test doesn't work if you haven't set the ancestor limit lower than the default.  It's in general tricky to test behavior which depends on the vagaries of coin selection, but one simple improvement to this test would be to send the coins to a different nodes address so then we only have the change to consider for subsequent spends and not the outputs themselves.  This would make it so you could reliably send (number of original outputs * chain limits) transactions as long as the change is large compared to the value you're sending.

However it would be nice if we could stop rebasing this PR and get it merged, so if you do want to change the test, please don't modify the other two commits...

**Comment by gmaxwell** [comment]
ACK.

File: src/wallet/wallet.cpp (hunk 14 above)
Comment (sipa) [inline src/wallet/wallet.cpp]: Given that we don't hold `mempool.cs` here, the transaction could be removed from the mempool in between the `pcoin->InMempool()` and `pcoin->GetMempoolAncestorCount()` calls, causing a segfault in the latter. Please abstract this whole conditional out into a method on `CTxMempool`, which takes the limit value and the txid, and returns whether the transaction is in the pool and satisfies the limits.

File: src/wallet/wallet.cpp [hunk 19]
```diff
@@ -1740,6 +1740,18 @@ bool CWalletTx::InMempool() const
...
+{
+    LOCK(mempool.cs);
+    return mempool.mapTx.find(GetHash())->GetCountWithAncestors();
+}
+
+uint64_t CWalletTx::GetMempoolDescendantCount() const
+{
+    LOCK(mempool.cs);
+    return mempool.mapTx.find(GetHash())->GetCountWithDescendants();
+}
```
Comment (sipa) [inline src/wallet/wallet.cpp]: Agree with @luke-jr, but I think we need something higher-level anyway (see further).

File: src/wallet/wallet.cpp [hunk 20]
```diff
@@ -2545,6 +2567,21 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
         }
     }
 
+    if (GetBoolArg("-walletrejectlongchains", DEFAULT_WALLET_REJECT_LONG_CHAINS)) {
+        // Lastly, ensure this tx will pass the mempool's chain limits
```
Comment (sipa) [inline src/wallet/wallet.cpp]: I don't care that it's called only once - this just doesn't seem like code that belongs in the wallet (which shouldn't need to know about the mempool's policy limits as much as possible). Having it abstracted probably makes it usable for other purposes too.

Also, #9290 does not mean we can't do belt and suspender checks to avoid producing long chains. Not adding such a transaction to the wallet (and mempool) is still superior to just adding it to the wallet and relying on retransmit to get it to go through at a later point in time.

File: src/wallet/wallet.h [hunk 21]
```diff
@@ -53,6 +53,8 @@ static const CAmount MIN_CHANGE = CENT;
 static const bool DEFAULT_SPEND_ZEROCONF_CHANGE = true;
 //! Default for -sendfreetransactions
 static const bool DEFAULT_SEND_FREE_TRANSACTIONS = false;
+//! Default for -walletrejectlongchains
+static const bool DEFAULT_WALLET_REJECT_LONG_CHAINS = false;
```
Comment (sipa) [inline src/wallet/wallet.h]: Why false?

**Comment by instagibbs** [comment]
@sipa Replaced SCMC mempool check with more generalized check in CTxMempool.

re: default I disagree, I think that the default behavior here makes more sense unless the user has a way of probing for something like `GetBalanceWithinChainLimits()`.

File: src/txmempool.h [hunk 22]
```diff
@@ -605,6 +605,9 @@ class CTxMemPool
     /** Expire all transaction (and their dependencies) in the mempool older than time. Return the number of removed transactions. */
     int Expire(int64_t time);
 
+    /** Returns false if the transaction is in the mempool and not within the chain limit specified. */
+    bool CoinWithinLimit(const uint256& txid, size_t chainLimit) const;
```
Comment (morcos) [inline src/txmempool.h]: Should we call this something other than **Coin**WithinLimit?  Seems like coins are not something that the mempool is aware of.

Comment (instagibbs) [inline src/txmempool.h]: `TransactionWithinChainLimit`?

File: src/wallet/wallet.h (hunk 21 above)
Comment (MarcoFalke) [inline src/wallet/wallet.h]: It is the least change in behavior, so makes sense in case this is still considered for backport.

File: src/txmempool.cpp [hunk 23]
```diff
@@ -1142,3 +1142,10 @@ void CTxMemPool::TrimToSize(size_t sizelimit, std::vector<uint256>* pvNoSpendsRe
     if (maxFeeRateRemoved > CFeeRate(0))
         LogPrint("mempool", "Removed %u txn, rolling minimum fee bumped to %s\n", nTxnRemoved, maxFeeRateRemoved.ToString());
 }
+
+bool CTxMemPool::TransactionWithinChainLimit(const uint256& txid, size_t chainLimit) const {
+    LOCK(cs);
+    if (exists(txid) && std::max(mapTx.find(txid)->GetCountWithAncestors(), mapTx.find(txid)->GetCountWithDescendants()) >= chainLimit)
```
Comment (sipa) [inline src/txmempool.cpp]: You're doing 3 map lookups here. What about
```c++
auto it = mapTx.find(txid);
return it != mapTx.end() && it->GetCountWithAncestors() < chainLimit &&
    it->GetCountWithDescendants() < chainLimit;
```

**Comment by sipa** [comment]
utACK 5882c099d9af6e8566d0bf46fb1da424a4373bf8, with one nit.

**Comment by instagibbs** [comment]
fixed @sipa nit

**Comment by laanwj** [comment]
Travis doesn't agree with your change, it seems.

**Comment by instagibbs** [comment]
@laanwj yes the logic made no sense, it will reject anything not in mempool rather than anything in mempool with too long a chain. Fixed.

File: src/wallet/wallet.cpp [hunk 24]
```diff
@@ -2545,6 +2555,21 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
...
+        LockPoints lp;
+        CTxMemPoolEntry entry(txNew, 0, 0, 0, 0, false, 0, false, 0, lp);
```
Comment (TheBlueMatt) [inline src/wallet/wallet.cpp]: Note that by not filling in the sigOpsCost here we dont have that information for the virtual transaction size, so you could still run over the limit here. I don't think its a big deal, but we should fix when we clean things up in 0.14 as suggested two lines up.

**Review by TheBlueMatt — APPROVED** [review]
utACK ff79329ccfe3ee4387b9de72d82431c7b8132272 some comments but none worth delaying 0.13.2 for.

File: src/txmempool.cpp [hunk 25]
```diff
@@ -1145,7 +1145,7 @@ void CTxMemPool::TrimToSize(size_t sizelimit, std::vector<uint256>* pvNoSpendsRe
 
 bool CTxMemPool::TransactionWithinChainLimit(const uint256& txid, size_t chainLimit) const {
     LOCK(cs);
-    if (exists(txid) && std::max(mapTx.find(txid)->GetCountWithAncestors(), mapTx.find(txid)->GetCountWithDescendants()) >= chainLimit)
-        return false;
-    return true;
+    auto it = mapTx.find(txid);
+    return it == mapTx.end() || (it->GetCountWithAncestors() < chainLimit &&
+    it->GetCountWithDescendants() < chainLimit);
```
Comment (gmaxwell) [inline src/txmempool.cpp]: This really should be indented.

**Review by btcdrak — APPROVED** [review]

**Comment by laanwj** [comment]
Backported in #9382, removing label
